### PR TITLE
bpo-36782: Created C API wrappers and added missing tests for functions in the PyDateTimeAPI.

### PR DESCRIPTION
**Pull request by edison12a**
https://bugs.python.org/issue36782

### CODE REVIEW

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]: Are these formatting changes deliberate? I know some editors will be somewhat aggressive with making formatting changes.

I think it would be best to leave the original formatting intact, if only to make the diff easier to review.

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]: Formatting nitpick: [PEP 7](https://www.python.org/dev/peps/pep-0007/#code-lay-out) says that for if statements, the curly brace goes on the line with the conditional. (The curly brace after the function definition *is* on its own line).

**Review by pganssle — CHANGES_REQUESTED** [review]
Thank you for this contribution! I'm very excited to have these added to the datetime test suite!

The Python portion of the code looks good to me, but I have some formatting notes on the C functions.

File: Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst
Comment (pganssle) [inline Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst]: Forgot to mention in my review, you can add "Patch by <yourname>." here.

**Comment by SimiCode** [comment]
I have made the changes that you requested. @pganssle Please review.

**Review by pganssle — APPROVED** [review]
Looks good to me! Thanks @SimiCode!

File: Lib/test/datetimetester.py
Comment (matrixise) [inline Lib/test/datetimetester.py]: maybe you could write 
```year, month, day = 1993, 8, 26```

File: Lib/test/datetimetester.py
Comment (matrixise) [inline Lib/test/datetimetester.py]: and because I don't like the duplicated code, you could use that.

```python
self.assertEqual(d, date(year, month, day))
```
I think you could also compute one time the expected date. and because I like to be explicit...

```python
def test_date_from_date(self):
    expected_date = date(1993, 8, 26)
    for macro in (0, 1):
        with self.subTest(macro=macro):
            new_date = _testcapi.get_date_fromdate(expected_date.year, expected_date.month, expected_date.day, macro)
            self.assertEqual(new_date, expected_date)
```

@pganssle ?

**Review by matrixise — CHANGES_REQUESTED** [review]
Thank you for your contribution, please read my comments with @pganssle and add you to the `Misc/ACKS` file.

**Comment by SimiCode** [comment]
I have made the requested changes; please review again

File: Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst
Comment (matrixise) [inline Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst]: @pganssle Could you help @SimiCode for the blurb entry? ```Add tests for the C-API wrapper of PyDate_FromDate ```? I am not a native speaker. or ```Add tests for PyDate_FromDate```

**Review by matrixise — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst
Comment (pganssle) [reply Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst]: Talked to Edison in the room, he will have an updated blurb that I have looked at. The current one is grammatically fine but could be tightened up and could use RST directives.

**Review by pganssle — APPROVED** [review]
New changes look good to me.

File: Misc/ACKS
Comment (matrixise) [inline Misc/ACKS]: And add a new line at end of file.

**Review by matrixise — COMMENTED** [review]
Thanks

**Comment by SimiCode** [comment]
I have made the requested changes; please review again.

**Comment by matrixise** [comment]
I can't merge your branch because you are working on this one. ping me when you will finish working on it. Thank you

**Comment by matrixise** [comment]
Ping me when you will finish working on it. Thank you. @pganssle You will need to review the updated PR of @SimiCode

File: Lib/test/datetimetester.py
Comment (pganssle) [inline Lib/test/datetimetester.py:6099]: Fold is not taken into account in datetime or time equality, you also need to assert that `c_api_time.fold == exp_time.fold`.

File: Lib/test/datetimetester.py
Comment (pganssle) [inline Lib/test/datetimetester.py:6068]: See the `time_fromtimeandfold` comment - need to check the fold explicitly.

File: Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst
Comment (pganssle) [inline Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst]:
```suggestion
Created tests for several C API functions in the datetime module, specifically::
```

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]: PEP 7 says "no line should be longer than 79 characters", so this and many other lines need to be broken. There are some rules on how to break them here: https://www.python.org/dev/peps/pep-0007/#code-lay-out

File: Lib/test/datetimetester.py
Comment (matrixise) [inline Lib/test/datetimetester.py]: add the `fold` parameter to `self.subTest` method

File: Lib/test/datetimetester.py
Comment (matrixise) [inline Lib/test/datetimetester.py]: ditto

File: Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst
Comment (matrixise) [inline Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst]: Just Add tests for C API functions. you don't need to specify the names of all the tested functions. One or two lines for the description is fine.

File: Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst
Comment (matrixise) [inline Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst]: `Patch by Edison Abahurire` is fine on the same line. no need to add a new blank line.

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c]: maybe too long, read PEP7 (https://www.python.org/dev/peps/pep-0007/)

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c]: see PEP7

**Review by matrixise — CHANGES_REQUESTED** [review]

**Review by pganssle — CHANGES_REQUESTED** [review]
A few changes here. I think the biggest thing is the PEP 7 compliance. I only mentioned the ones I noticed, but it's hard to keep all the formatting rules in my head...

**Comment by SimiCode** [comment]
I have made the changes that you requested.
To comply with PEP7, I grouped together variables that are similar on new lines:
```
    if (!PyArg_ParseTuple(
        args, 
        "iiiiiiii|p",
        &year, &month, &day,
        &hour, &minute, &second, &microsecond,
        &fold,
        &macro)) {

        return NULL;
    }
```

Is this OK?

File: Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst
Comment (matrixise) [inline Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst]: just remove `::`

Comment (matrixise) [reply Misc/NEWS.d/next/Tests/2019-05-04-21-25-19.bpo-36782.h3oPIb.rst]: maybe this sentence `Add tests for several C API functions in the :mod:`datetime` module. Patch by Edison Abahurire`

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c]: remove the blank line

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c:2372]:
```c++
if (!PyArg_ParseTuple(args, "iiiiiii|p",
                      &year, &month, &day, &hour, &minute, &second, &microsecond,
                      &macro)) {
    return NULL;
}
```

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c]: ditto, 
remove empty string and reformat your call to `PyArg_ParseTuple`
from PEP7 (https://www.python.org/dev/peps/pep-0007/)

Breaking long lines: if you can, break after commas in the outermost argument list. Always indent continuation lines appropriately, e.g.:

in this case, 
```c++
if (!PyArg_ParseTuple(args, "iiiiiiii|p",
                      &year, &month, &day,
                      &hour, &minute, &second, &microsecond,
                      &fold, &macro)) {
    return NULL;
}
```

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c:2412]: same here

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c:2418]: same here

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c]: same here

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c]: same here

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c]: same here

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c:2497]: Breaking long lines: if you can, break after commas in the outermost argument list. Always indent continuation lines appropriately, e.g.:

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c:2471]: just reformat, ```Breaking long lines: if you can, break after commas in the outermost argument list. Always indent continuation lines appropriately, e.g.:```

File: Modules/_testcapimodule.c
Comment (matrixise) [inline Modules/_testcapimodule.c]: this curly brace has to be on the first column

**Review by matrixise — CHANGES_REQUESTED** [review]

File: Modules/_testcapimodule.c
Comment (vstinner) [inline Modules/_testcapimodule.c]: I would prefer to make macro mandatory. Maybe it should be moved at the beginning, but I would prefer to get Paul's call on that.

File: Modules/_testcapimodule.c
Comment (vstinner) [inline Modules/_testcapimodule.c]: Since these parameters are mandatory, there is no need to initialize them.

File: Modules/_testcapimodule.c
Comment (vstinner) [inline Modules/_testcapimodule.c]: This comment looks useless to me, please remove it.

File: Modules/_testcapimodule.c
Comment (vstinner) [inline Modules/_testcapimodule.c]: nitpick: I suggest this empty line ;-) It doesn't add any value in term of readability.

My comments apply to all C functions below.

**Review by vstinner — CHANGES_REQUESTED** [review]
I like the overall approach, but I have questions/comments :-)

@pganssle: Is "PyDateTime_IMPORT;" needed somewhere? What if only one test method is called? Magic PyDateTimeAPI static global is initialized to NULL, no?

@pganssle: I would prefer to put "macro" parameter at the start rather than at the end. What do you think?

**Comment by vstinner** [comment]
@pganssle: Would you mind to review this PR again? @SimiCode updated it. Once you approve it, I'm ok to review it to maybe merge it ;-)

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]:
```suggestion
    int macro;
```

If we're making this required, it does not have to be initialized.

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]:
```suggestion
    int macro;
```

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]:
```suggestion
    int macro;
```

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]:
```suggestion
    int macro;
```

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]:
```suggestion
    int macro;
```

File: Modules/_testcapimodule.c
Comment (pganssle) [inline Modules/_testcapimodule.c]:
```suggestion
    int macro;
```

**Review by pganssle — APPROVED** [review]
I've suggested changing all the `int macro = 0` to `int macro;` now that `macro` is required, but otherwise this PR looks good and we can merge it.

**Comment by SimiCode** [comment]
@vstinner @pganssle 
I have added the suggested changes. Please review.

**Review by matrixise — APPROVED** [review]
Thank you for your contribution. 🎉

**Review by pganssle — APPROVED** [review]
LGTM! Thanks Edison!

**Comment by vstinner** [comment]
Well done @SimiCode, modifying C code and tests is not trivial!

**Review by vstinner — APPROVED** [review]

**Comment by vstinner** [comment]
Crap. I modified the commit message, but the merge failed. I click on Retry, and then it lost all my changes on the commit message :-(

**Comment by vstinner** [comment]
@pganssle: Do we need to backport the change to 3.7?

**Comment by pganssle** [comment]
@vstinner It only adds tests, so while I don't think it would *hurt* to backport the change to 3.7, it's not really necessary.

**Pull request merged**
